### DISCUSSION TITLE
zh/integrations/tensorrt/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9868] [Docs]
# zh/integrations/tensorrt/

探索将Ultralytics YOLOv8  模型导出为TensorRT 格式的强大功能和灵活性，以提高在英伟达™（NVIDIA®）图形处理器上的性能和效率。

https://docs.ultralytics.com/zh/integrations/tensorrt/

**Comment by abcc235** [comment]
请问在那儿可以看到
results = tensorrt_model('https://ultralytics.com/images/bus.jpg')中
results的数据结构。

**Reply by glenn-jocher** [reply]
您好！关于`results`的数据结构，它包含了模型推理的输出，通常包括：

- **检测框** (`boxes`): 每个检测对象的位置和大小，通常以`[x_min, y_min, x_max, y_max]`的形式表示。
- **置信度** (`confidences`): 模型对每个检测对象的置信度。
- **类别ID** (`classIDs`): 每个检测对象的预测类别ID。

当您运行如下代码:

```python
results = tensorrt_model('https://ultralytics.com/images/bus.jpg')
```

您可以通过`results.xyxy`来获取**检测框**，通过`results.conf`获取**置信度**，通过`results.cls`来获取**类别ID**。例如，打印所有检测框及其对应的置信度和类别ID：

```python
# 获取xyxy格式的检测框（x_min, y_min, x_max, y_max），置信度和类别ID
for box, conf, cls_id in zip(results.xyxy[0], results.conf[0], results.cls[0]):
    print('检测框:', box.cpu().numpy(), '置信度:', conf, '类别ID:', int(cls_id))
```
希望这能帮助到您！🙂

**Reply by abcc235** [reply]
貌似应该这么写，直接按照您写的会报错。
、、、python
        for box, conf, cls_id in zip(results[0].boxes.xyxy, results[0].boxes.conf, results[0].boxes.cls):
            print('检测框:', box.cpu().numpy(), '置信度:', conf, '类别ID:', int(cls_id))
、、、

**Reply by glenn-jocher** [reply]
@abcc235 您好！看起来您已经正确地修改了访问`Results`对象属性的方法，确实需要适当调用属性。您的代码片段中提到将`results`的索引直接解包到循环中，这是非常合适的做法。以下是一个简化的代码示例，确保您可以顺利读取和打印出检测结果：

```python
results = tensorrt_model('https://ultralytics.com/images/bus.jpg')
for box, conf, cls_id in zip(results[0].boxes.xyxy, results[0].boxes.conf, results[0].boxes.cls):
    print('检测框:', box.cpu().numpy(), '置信度:', conf.item(), '类别ID:', cls_id.item())
```

这里使用`.item()`确保从Tensor中提取标量值，使输出更加清晰。如果有进一步问题或需求，欢迎随时提问！🙂

**Comment by 17638759823** [comment]
(yolov8) PS E:\ultralytics> yolo export model=./weights/yolov8n.pt format=engine                          
WARNING  TensorRT requires GPU export, automatically assigning device=0
Ultralytics YOLOv8.0.199  Python-3.8.18 torch-2.1.0+cu121 CUDA:0 (NVIDIA GeForce RTX 3050 OEM, 8192MiB)
YOLOv8n summary (fused): 168 layers, 3151904 parameters, 0 gradients, 8.7 GFLOPs

PyTorch: starting from 'weights\yolov8n.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (6.2 MB)

ONNX: starting export with onnx 1.15.0 opset 17...
到这就停止了，这是什么问题

**Reply by pderrenger** [reply]
看起来您在导出YOLOv8模型为TensorRT格式时遇到了中断。通常，这个问题可能与几个不同因素相关，包括TensorRT的安装问题、GPU兼容性，或者可能是特定版本的PyTorch或ONNX与TensorRT之间的兼容性问题。

首先，请确保您已正确安装所有必需的TensorRT依赖项，以及NVIDIA驱动和CUDA版本与您的TensorRT版本兼容。此外，建议检查您的PyTorch版本是否支持您正在尝试导出的ONNX版本。

此外，试着使用以下导出命令，并确保`device`被设定为可用的GPU：
```bash
yolo export model=./weights/yolov8n.pt format=engine device=0
```
如果问题仍然存在，请尝试更新到最新版本的PyTorch和TensorRT，或考虑在Ultralytics的GitHub问题跟踪器中查找类似的已知问题或提交新的问题。确保提供您的操作系统、PyTorch、CUDA和TensorRT版本的详细信息，以便社区更好地帮助您解决问题。🤖

有关更多细节和最佳实践，强烈建议阅读我们的[TensorRT导出指南](https://docs.ultralytics.com/zh/integrations/tensorrt/)，以获取从YOLOv8模型到TensorRT模型的转换步骤、安装要求和示例代码。希望这能帮到您！

**Comment by Ares-cz** [comment]
when I use official demo to export into TensorRT, it gave me a feedback on the following error:

AttributeError: 'tensorrt_bindings.tensorrt.IBuilderConfig' object has no attribute 'max_workspace_size'
[06/30/2024-14:20:22] [TRT] [I] [MemUsageChange] Init builder kernel library: CPU +2514, GPU +288, now: CPU 14191, GPU 1967 (MiB)
TensorRT: export failure ❌ 2.1s: 'tensorrt_bindings.tensorrt.IBuilderConfig' object has no attribute 'max_workspace_size'

the system is Windows 11, and the CUDA version is 12.1, the TensorRT version is 10.1

**Reply by glenn-jocher** [reply]
@Ares-cz hello,

Thank you for reaching out and providing details about the issue you're encountering. The error you're seeing, `AttributeError: 'tensorrt_bindings.tensorrt.IBuilderConfig' object has no attribute 'max_workspace_size'`, suggests there might be a compatibility issue with the versions of TensorRT or CUDA you are using.

Here are a few steps to help resolve this issue:

1. **Update Packages**: Ensure you are using the latest versions of all relevant packages, including `ultralytics`, `tensorrt`, and `CUDA`. You can update `ultralytics` using:
   ```bash
   pip install --upgrade ultralytics
   ```

2. **Reproducible Example**: If the issue persists, it would be very helpful if you could provide a minimum reproducible example. This will allow us to better understand the context and provide a more accurate solution. You can find guidelines on creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

3. **Export Code**: Ensure your export code follows the recommended format. Here is an example:
   ```python
   from ultralytics import YOLO

   # Load the YOLOv8 model
   model = YOLO("yolov8n.pt")

   # Export the model to TensorRT format
   model.export(format="engine")  # creates 'yolov8n.engine'
   ```

4. **Check TensorRT Documentation**: For more detailed instructions and best practices, please refer to the [TensorRT Export Guide](https://docs.ultralytics.com/integrations/tensorrt/).

If you continue to experience issues, please share the specific versions of `ultralytics`, `tensorrt`, and `CUDA` you are using, along with any additional error messages. This information will help us diagnose the problem more effectively.

Thank you for your cooperation, and we look forward to helping you resolve this issue! 😊

**Comment by pdw2002** [comment]
您好，我在使用yolov8.pt权重转换成engine格式进行测试时，发现engine还没有原先的.pt快，这是两者相比较的速度
best-seg.pt
imgsz=2560,device='3,batch=16
Speed: 39.6ms preprocess, 20.7ms inference, 1.9ms postprocess per image at shape (16, 3, 1440, 2560)

best-seg.engine
imgsz=2560,device='3',batch=16
Speed: 62.3ms preprocess, 26.8ms inference, 1.9ms postprocess per image at shape (16, 3, 2560, 2560)

其它参数都是一样的
这是转换engine的代码：
from ultralytics import YOLO

model = YOLO(r"best-seg.pt")  # load a custom trained model

# Export the model
model.export(
    task="segment",
    format="engine",
    imgsz=(2560,2560),
    batch=16,
    device='0',
    simplify=True,
    int8=True,
             )


这是推理代码：
from ultralytics import YOLO
if __name__ == '__main__':
    # 加载模型
    model = YOLO(r'best-seg.engine')  # YOLOv8n模型
    results = model.predict(
        # source=r'ultralytics-8.2.69\ultralytics\assets\test_data',
        source=r'test.MP4',
        save=True,  # 保存预测结果
        imgsz=2560,  # 输入图像的大小，可以是整数或w，h
        conf=0.2,  # 用于检测的目标置信度阈值（默认为0.25，用于预测，0.001用于验证）
        iou=0.5,  # 非极大值抑制 (NMS) 的交并比 (IoU) 阈值
        batch=16,  # 设置多GPU推理时需要指明批次大小，而不能使用自动批量大小
        device="3",
        show=False,  # 如果可能的话，显示结果
        project='runs/predict',  # 项目名称（可选）
        name='exp',  # 实验名称，结果保存在'project/name'目录下（可选）
        save_txt=True,  # 保存结果为 .txt 文件
        save_conf=True,  # 保存结果和置信度分数
        save_crop=True,  # 保存裁剪后的图像和结果
        show_labels=True,  # 在图中显示目标标签
        show_conf=True,  # 在图中显示目标置信度分数
        vid_stride=1,  # 视频帧率步长
        line_width=1,  # 边界框线条粗细（像素）
        visualize=False,  # 可视化模型特征
        augment=False,  # 对预测源应用图像增强
        agnostic_nms=False,  # 类别无关的NMS
        retina_masks=False,  # 使用高分辨率的分割掩码
        show_boxes=True,  # 在分割预测中显示边界框
        stream=True,  # (True)结果以生成器形式返回，逐步处理，适合大批量数据处理
        stream_buffer=False,  # 缓冲所有流帧（True）或返回最新帧（False）
    )

    # 处理结果
    for result in results:
        if result is not None:
            # 使用result.path属性作为文件名的一部分
            result.save(filename=f"result_{result.path.split('/')[-1]}")
        else:
            print("No detections in this frame")

我的疑惑是也调用了GPU进行加速推理，然后推理结果反而比原来的慢，我不清楚是什么原因，请解答我的疑惑，谢谢。

**Reply by pdw2002** [reply]
补充一下，GPU是RTX-2080-Ti

**Reply by glenn-jocher** [reply]
@pdw2002 您好，感谢您的反馈。

请确保您使用的是最新版本的Ultralytics库，并尝试在导出时不使用`int8=True`选项，因为INT8量化可能会影响推理速度。此外，您可以尝试调整`workspace`参数以优化性能。如果问题仍然存在，请访问我们的[TensorRT文档](https://docs.ultralytics.com/integrations/tensorrt/)获取更多信息和指导。

希望这能帮助您解决问题。谢谢！

**Comment by NidhoggBo** [comment]
AttributeError: 'tensorrt_bindings.tensorrt.ICudaEngine' object has no attribute 'num_bindings'

# Load the exported TensorRT model
tensorrt_model = YOLO("yolov8n.engine")

# Run inference
results = tensorrt_model("https://ultralytics.com/images/bus.jpg")

**Reply by glenn-jocher** [reply]
@NidhoggBo the error `'ICudaEngine' object has no attribute 'num_bindings'` typically occurs due to an incompatibility between the TensorRT version used during export and the runtime environment. Please ensure that your TensorRT version is compatible with the exported engine. Additionally, verify that the model was exported correctly using the `model.export(format="engine")` command, as shown in the [Ultralytics TensorRT guide](https://docs.ultralytics.com/integrations/tensorrt/). Let us know if further assistance is needed.

**Comment by Ameridru** [comment]
Ultralytics是否支持对RT-DETR的TensorRT模型进行export以及predict？

**Reply by glenn-jocher** [reply]
@Ameridru yes, Ultralytics supports exporting RT-DETR models to TensorRT format and running predictions. You can use the standard export workflow:

```python
from ultralytics import RTDETR

model = RTDETR('rtdetr-l.pt')  # Load model
model.export(format='engine')  # Export to TensorRT
results = model.predict('image.jpg')  # Run inference
```

For deployment details, see our RT-DETR documentation at https://docs.ultralytics.com/models/rtdetr.

**Comment by Hammer-888** [comment]
I got low acc by export yolov5n.engine, do we need pics to correction int8 model

**Reply by glenn-jocher** [reply]
@Hammer-888 for INT8 quantization with TensorRT exports, calibration images from your dataset are required. Ensure you're using the `data` argument pointing to your dataset YAML (like `data=coco.yaml`) during export to enable proper calibration, as shown in the TensorRT INT8 export examples. This helps minimize accuracy loss while maintaining performance gains.

**Comment by AK993** [comment]
这个方法适用于rt-detr吗

**Comment by fly2skyToEnjoy** [comment]
May I ask : The following command exports the file, whether the box and mask are both nmsed at the same time ???



from ultralytics import YOLO

# Load the YOLO11 model
model = YOLO("yolo11n.pt")

# Export the model to TensorRT format
model.export(format="engine")  # creates 'yolo11n.engine'

# Load the exported TensorRT model
tensorrt_model = YOLO("yolo11n.engine")

# Run inference
results = tensorrt_model("https://ultralytics.com/images/bus.jpg")

**Comment by 020423lzl** [comment]
etson@yahboom:~/ultralytics/ultralytics$ yolo predict model=best.engine source=0 save=False show
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Ultralytics 8.3.172 🚀 Python-3.10.12 torch-2.5.0a0+872d972e41.nv24.08 CUDA:0 (Orin, 7620MiB)
Loading best.engine for TensorRT inference...
[12/18/2025-17:11:05] [TRT] [I] Loaded engine size: 75 MiB
[12/18/2025-17:11:05] [TRT] [I] [MS] Running engine with multi stream info
[12/18/2025-17:11:05] [TRT] [I] [MS] Number of aux streams is 7
[12/18/2025-17:11:05] [TRT] [I] [MS] Number of total worker streams is 8
[12/18/2025-17:11:05] [TRT] [I] [MS] The main stream provided by execute/enqueue calls is the first worker stream
[12/18/2025-17:11:05] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +1, GPU +88, now: CPU 1, GPU 158 (MiB)

[ WARN:0@2.771] global cap_gstreamer.cpp:1777 open OpenCV | GStreamer warning: Cannot query video position: status=0, value=-1, duration=-1
1/1: 0... Success ✅ (inf frames of shape 1920x1080 at 5.00 FPS)

Traceback (most recent call last):
  File "/home/jetson/.local/bin/yolo", line 7, in <module>
    sys.exit(entrypoint())
  File "/home/jetson/Downloads/ultralytics-main/ultralytics/cfg/__init__.py", line 983, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/home/jetson/Downloads/ultralytics-main/ultralytics/engine/model.py", line 555, in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
  File "/home/jetson/Downloads/ultralytics-main/ultralytics/engine/predictor.py", line 247, in predict_cli
    for _ in gen:  # sourcery skip: remove-empty-nested-block, noqa
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 36, in generator_context
    response = gen.send(None)
  File "/home/jetson/Downloads/ultralytics-main/ultralytics/engine/predictor.py", line 351, in stream_inference
    s[i] += self.write_results(i, Path(paths[i]), im, s)
  File "/home/jetson/Downloads/ultralytics-main/ultralytics/engine/predictor.py", line 434, in write_results
    string += f"{result.verbose()}{result.speed['inference']:.1f}ms"
  File "/home/jetson/Downloads/ultralytics-main/ultralytics/engine/results.py", line 694, in verbose
    return "".join(f"{n} {self.names[i]}{'s' * (n > 1)}, " for i, n in enumerate(counts) if n > 0)
  File "/home/jetson/Downloads/ultralytics-main/ultralytics/engine/results.py", line 694, in <genexpr>
    return "".join(f"{n} {self.names[i]}{'s' * (n > 1)}, " for i, n in enumerate(counts) if n > 0)
KeyError: 171
terminate called without an active exception
Aborted (core dumped)
jetson@yahboom:~/ultralytics/ultralytics$ How to solve it

**Reply by glenn-jocher** [reply]
`KeyError: 171` 表示模型输出了类别 `171`，但当前 `names` 里没有这个索引（通常是自定义数据集的标签类别编号不在 `0..nc-1` 范围内、`data.yaml` 里的 `names` 数量与 `nc` 不匹配，或导出/推理用的模型与数据配置不一致导致的元数据错配）；请先把所有标注 `.txt` 的 class id 修正为连续的 `0..nc-1`，同步修正 `data.yaml` 的 `nc`/`names`，修正后删除数据集 `labels/` 下的 `*.cache`，再用同一套环境从 `best.pt` 重新导出 `.engine`（参考 [TensorRT 导出文档](https://docs.ultralytics.com/zh/integrations/tensorrt/)），并在推理时显式加上 `task=detect`。另外请把下面命令的完整输出（文本）贴在这个 discussion 里，以及你当时的导出命令，方便我们确认是否有多环境/路径冲突：  
```bash
yolo checks
```

**Reply by 020423lzl** [reply]
Thank you for your reply. The issue has been resolved. I was using the RT-DETR model, but I mistakenly used "from ultralytics import YOLO" when exporting

**Reply by glenn-jocher** [reply]
Glad you found it — RT-DETR models should be loaded/exported with `RTDETR(...)` (mixing them with `YOLO(...)` can produce mismatched metadata like `names`), see the [RT-DETR model docs](https://docs.ultralytics.com/models/rtdetr/) for details:

```python
from ultralytics import RTDETR

model = RTDETR("best.pt")
model.export(format="engine")  # creates best.engine
```

**Comment by johnfans** [comment]
现在有一个奇怪的现象，我导出的tensorrt模型的推理速度几乎和原来的.pt几乎没有差别甚至更慢，并且导出的tensorrt模型的推理结果和.pt模型的差别很大。
我的设备是i914900k+64G RAM+RTX A6000
我的环境是cuda12.4.131+tensorrt10.10.0.31+torch2.5.1+ultralytics8.3.252
导出命令：
```yolo export model=yolo11m.pt imgsz=1280 format='engine```
推理代码为：
```python
model.predict(stream_path, imgsz=1280, classes=[14], conf=0.2, iou=0.5, stream=True, task='detect')
```
另外我试了下yolo11n.pt的640输入尺寸导出，推理速度就有明显提升，当然推理结果差别仍然较大。
请问这是什么道理，有什么解决办法？

**Comment by 020423lzl** [comment]
m的模型大小比n大的多，速度也慢了很多，输入图像尺寸越大，越吃显存，速度也越慢












李子乐控制工程
***@***.***






---Original---
From: ***@***.***&gt;
Date: Sat, May 9, 2026 10:30 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ultralytics/ultralytics] zh/integrations/tensorrt/ (Discussion #9868)




 
现在有一个奇怪的现象，我导出的tensorrt模型的推理速度几乎和原来的.pt几乎没有差别甚至更慢，并且导出的tensorrt模型的推理结果和.pt模型的差别很大。
 我的设备是i914900k+64G RAM+RTX A6000
 我的环境是cuda12.4.131+tensorrt10.10.0.31+torch2.5.1+ultralytics8.3.252
 导出命令：
 yolo export model=yolo11m.pt imgsz=1280 format='engine
 推理代码为：
 model.predict(stream_path, imgsz=1280, classes=[14], conf=0.2, iou=0.5, stream=True, task='detect')
 
另外我试了下yolo11n.pt的640输入尺寸导出，推理速度就有明显提升，当然推理结果差别仍然较大。
 请问这是什么道理，有什么解决办法？
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by johnfans** [reply]
能不能读完再说话，肯定是m和m比，n和n比，同样尺寸输入和同样尺寸输入的比啊。
该显卡的RT core算力是普通fp32算力的近两倍。

**Reply by glenn-jocher** [reply]
是的，性能应当在同一模型、同一 `imgsz`、同一 `batch`、同一后处理参数下比较；`m` 比 `n` 更大、`1280` 比 `640` 更吃显存且更慢是正常现象。对 RTX A6000 来说，`FP32` 的 TensorRT 往往不会比 `.pt` 明显更快，建议先按 [TensorRT 导出文档](https://docs.ultralytics.com/zh/integrations/tensorrt/) 用 `half=True` 重新导出，并在相同输入上预热后再统计平均耗时；如果 `engine` 与 `.pt` 结果仍明显不一致，请贴出 `!yolo checks`、完整控制台日志文本和一个最小可复现示例，我们好继续排查环境或后处理差异。
